src/Features/CSharp/Portable/CodeRefactorings/ConvertToRecord/PropertyAnalysisResult.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.ConvertToRecord
+{
+    internal record PropertyAnalysisResult(PropertyDeclarationSyntax Syntax, IPropertySymbol Symbol, bool KeepAsOverride)
+    {
+
+        public static ImmutableArray<PropertyAnalysisResult> AnalyzeProperties(
+            ImmutableArray<PropertyDeclarationSyntax> properties,
+            INamedTypeSymbol type,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            // first get all property symbols
+            var symbols = properties
+                .SelectAsArray(p => (IPropertySymbol)semanticModel.GetRequiredDeclaredSymbol(p, cancellationToken));
+
+            // The user may not know about init or be converting code from before init was introduced.
+            // In this case we can convert set properties to init ones
+            var allowSetToInitConversion = !symbols
+                .Any(symbol => symbol.SetMethod is IMethodSymbol { IsInitOnly: true });
+
+            return properties.ZipAsArray(symbols, (syntax, symbol)
+                => ShouldConvertProperty(syntax, symbol, type) switch
+                {
+                    ConvertStatus.DoNotConvert => null,
+                    ConvertStatus.Override => new PropertyAnalysisResult(syntax, symbol, true),
+                    ConvertStatus.OverrideIfConvertingSetToInit
+                        => new PropertyAnalysisResult(syntax, symbol, !allowSetToInitConversion),
+                    ConvertStatus.AlwaysConvert => new PropertyAnalysisResult(syntax, symbol, false),
+                    _ => throw ExceptionUtilities.Unreachable,
+                }).WhereNotNull().AsImmutable();
+        }
+
+        // for each property, say whether we can convert
+        // to primary constructor parameter or not (and whether it would imply changes)
+        private enum ConvertStatus
```

Comment:
The main point of the enum is to determine whether we should leave the property declaration in as an override. For example, if we see that the user uses only `get; set;` on all their properties, it makes sense that they might just want to turn them into `get; init;` since they might have written this code before `init` was introduced. However, if they have some properties as `get; set;` and some as `get; init;`, then we would want to provide an explicit override to the `set` properties, since the user clearly wants them to be `set` and not `init`